opencog/atoms/distvalue/DistributionalValue.cc

```diff
@@ -0,0 +1,455 @@
+{
+	DVec res;
+	for (auto interval : k) {
+		if (interval.size() == 1)
+			res.push_back(interval[0]);
+		res.push_back((interval[0] + interval[1])/2);
+	}
+	return res;
+}
+
+//Get the variance for all Keys
+std::vector<double> DistributionalValue::get_var() const
+{
+	std::vector<double> probs;
+	for (auto elem : _value)
+	{
+		probs.push_back(get_var_for(elem.second));
+	}
+	return probs;
+}
+
+//Get the variance for a Count
+double DistributionalValue::get_var_for(double ai) const
+{
+		double a0 = total_count();
+		return ai*(a0-ai) / a0*a0*(a0+1);
+}
+
+double DistributionalValue::total_count() const
+{
+	return _value.total_count();
+}
+
+double DistributionalValue::get_confidence() const
+{
+	int c = total_count();
+	return to_conf(c);
+}
+
+double DistributionalValue::to_conf(int c)
+{
+	return c / (c + DEFAULT_K);
+}
+
+int DistributionalValue::to_count(double cf)
+{
+	return (cf * DEFAULT_K / (1 - cf));
+}
+
+bool DistributionalValue::has_key(const DVKey &k) const
+{
+	auto it = _value.find(k);
+	return it != _value.end();
+}
+
+DVKeySeq DistributionalValue::get_keys() const
+{
+	DVKeySeq res;
+	for (auto k : _value)
+		res.push_back(k.first);
+	return res;
+}
+
+double DistributionalValue::get_count(const DVKey &h) const
+{
+	auto pos = _value.find(h);
+	if (pos != _value.end())
+		return pos->second;
+	throw RuntimeException(TRACE_INFO, "No Key for this value.");
+}
+
+//Find out how much of Key1 is contained in Key2
+double DistributionalValue::key_contained(const DVKey &ks1,const DVKey &ks2)
+{
+	Interval k1,k2;
+	//Start with the assumption that 100% of Key1 is in Key2
+	double sum = 1;
+	for (auto zipped : boost::combine(ks1,ks2))
+	{
+		boost::tie(k1,k2) = zipped;
+		//Singleton Intervals are equal => assumption is correct and we can continue
+		if (k1.size() == 1 && 1 == k2.size() && k1[0] == k2[0])
+			continue;
+
+		//Singleton Intervals are different => no overlap return 0
+		if (k1.size() == 1 && 1 == k2.size() && k1[0] != k2[0])
+			return 0;
+
+		//Singleton Interval of Key1 is contained completely in Interval of Key2
+		if (k1.size() == 1 && 2 == k2.size() && k1[0] > k2[0] && k1[0] < k2[1])
+			continue;
+
+		//Singleton Interval of Key1 isn't contained in Interval of Key2
+		if (k1.size() == 1 && 2 == k2.size() && (k1[0] < k2[0] || k1[0] > k2[1]))
+			return 0;
+
+		// The part of an interval contained in a sigelton set is infinitelly small
+		if (k1.size() == 2 && 1 == k2.size())
+			return 0;
+
+		//Two Intervals
+		//Calculate the overlapp
+		Interval res;
+		//Find the start of the Overlapp
+		if (k1[0] > k2[0])
+			res.push_back(k1[0]);
+		else
+			res.push_back(k2[0]);
+
+		//Find the end of the Overlapp
+		if (k1[1] < k2[1])
+			res.push_back(k1[1]);
+		else
+			res.push_back(k2[1]);
+
+		//Check that the Overlapp starts before it ends
+		if (res[0] <= res[1])
+			//Figure out how much smaler the Overlapp is then K1
+			sum *= (res[1] - res[0]) / (k1[1] - k1[0]);
+		else
+			return 0;
```

Comment:
That looks like a Jaccard index to me, maybe that's how it should be called. There's BTW a `jaccard_index` function in cogutil but it's for sets, not intervals.

Reply:
Sorry, it's not the Jaccard index. It must have a standard name though...

Reply:
It's basically a conditional probability, maybe that's how it should be called...

Reply:
Oh, you are right the intervals are basically continuous uniform distributions this calculates the conditional probability. Going to rename it and add some comments.
Would be nice if we could use other distributions but I don't know of a library that implements that nicely. std/boost have distributions but they are mainly for generating random numbers not to calculate conditional probabilities between them.

Reply:
Maybe have a look at https://www.boost.org/doc/libs/1_69_0/libs/math/doc/html/dist.html

The URE code uses https://www.boost.org/doc/libs/1_69_0/libs/math/doc/html/math_toolkit/dist_ref/dists/beta_dist.html to calculate beta distribution pdf and other things.

Though I suppose uniform distribution is OK for starter.

Reply:
hm, I guess using the pdf and some integration I could calculate the conditional probability on my own.

Reply:
Boost can handle the cdf as well.